pkg/reconciler/events/pubsub/pubsub.go

```diff
@@ -63,89 +60,99 @@ type Reconciler struct {
 	if apierrs.IsNotFound(err) {
 		// The PubSub resource may no longer exist, in which case we stop processing.
-		runtime.HandleError(fmt.Errorf("storage '%s' in work queue no longer exists", key))
+		logging.FromContext(ctx).Desugar().Error("PubSub in work queue no longer exists")
 		return nil
 	} else if err != nil {
 		return err
 	}
 
 	// Don't modify the informers copy
-	csr := original.DeepCopy()
+	pubsub := original.DeepCopy()
+
+	reconcileErr := r.reconcile(ctx, pubsub)
 
-	reconcileErr := r.reconcile(ctx, csr)
+	// If no error is returned, mark the observed generation.
+	if reconcileErr == nil {
+		pubsub.Status.ObservedGeneration = pubsub.Generation
+	}
 
-	if equality.Semantic.DeepEqual(original.Status, csr.Status) &&
-		equality.Semantic.DeepEqual(original.ObjectMeta, csr.ObjectMeta) {
+	if equality.Semantic.DeepEqual(original.Status, pubsub.Status) {
 		// If we didn't change anything then don't call updateStatus.
 		// This is important because the copy we loaded from the informer's
 		// cache may be stale and we don't want to overwrite a prior update
 		// to status with this stale state.
-	} else if _, err := r.updateStatus(ctx, csr); err != nil {
-		// TODO: record the event (c.Recorder.Eventf(...
-		r.Logger.Warn("Failed to update PubSub Source status", zap.Error(err))
-		return err
-	}
 
+	} else if _, uErr := r.updateStatus(ctx, pubsub); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update PubSub status", zap.Error(uErr))
+		r.Recorder.Eventf(pubsub, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for PubSub %q: %v", pubsub.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(pubsub, corev1.EventTypeNormal, "Updated", "Updated PubSub %q", pubsub.Name)
+	}
 	if reconcileErr != nil {
-		// TODO: record the event (c.Recorder.Eventf(...
-		return reconcileErr
+		r.Recorder.Event(pubsub, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
 	}
-
-	return nil
+	return reconcileErr
 }
 
-func (r *Reconciler) reconcile(ctx context.Context, source *v1alpha1.PubSub) error {
-	source.Status.ObservedGeneration = source.Generation
-	source.Status.InitializeConditions()
+func (r *Reconciler) reconcile(ctx context.Context, pubsub *v1alpha1.PubSub) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("pubsub", pubsub)))
+
+	pubsub.Status.InitializeConditions()
 
-	if source.DeletionTimestamp != nil {
+	if pubsub.DeletionTimestamp != nil {
 		// No finalizer needed, the pullsubscription will be garbage collected.
 		return nil
 	}
 
-	ps, err := r.reconcilePullSubscription(ctx, source)
+	ps, err := r.reconcilePullSubscription(ctx, pubsub)
 	if err != nil {
-		r.Logger.Infof("Failed to reconcile PubSub: %s", err)
+		pubsub.Status.MarkPullSubscriptionNotReady("PullSubscriptionCreateFailed", "Failed to reconcile PullSubscription: %s", err.Error())
```

Comment:
Why is the reason `PullSubscriptionCreateFailed`? Can this error only happen on create?

Reply:
I guess the correct term would be `PullSubscriptionReconcileFailed`. It can happen during a reconciliation (e.g., when getting the resource, creating, in some cases updating, etc.)
I actually renamed most methods in the reconcilers to `reconcile*`, they used to be `create*`, and marked them `*ReconcileFailed` in case of errors. 
I later on went back to `*CreateFailed`, as IMO was more intuitive for users. My thought was that many folks don't know what a reconciliation is. 
I can go back to `*ReconcileFailed` everywhere if you prefer. I don't have a strong preference.

Reply:
I don't have a strong preference either, but `CreateFailed` seems wrong when the resource has already been created and the error is `get failed`. Not a big deal, can be a followup.

Reply:
You are both right, finer-grained reasons would be better. And `CreateFailed` might be wrong. 
Let me go back to `ReconcileFailed` because it's an easier change than adding the finer-grain reasons. Otherwise I'll have to make changes in a bunch of places.
We can add those finer-grained reasons later on. I think with the detailed error message might be enough.

Reply:
Done!